lms/services/digest/service.py

```diff
@@ -0,0 +1,109 @@
+from collections import defaultdict
+from datetime import datetime
+from typing import Iterator, List
+
+from lms.services.digest._digest_assistant import DigestAssistant
+from lms.services.digest._models import Digest
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
+
+
+class DigestService:
+    """A service for getting digests and emailing them to people."""
+
+    def __init__(
+        self,
+        digest_assistant: DigestAssistant,
+        h_api: HAPI,
+        mailchimp_service: MailchimpService,
+        sender: EmailSender,
+    ):
+        self._assistant = digest_assistant
+        self._h_api = h_api
+        self._mailchimp_service = mailchimp_service
+        self._sender = sender
+
+    def send_emails(
+        self,
+        audience: List[str],
+        updated_after: datetime,
+        updated_before: datetime,
+        override_to_email: str = None,
+    ):
+        """
+        Prepare and send activity digests for users over the given period.
+
+        :param audience: List of H userids to send digests to
+        :param updated_after: Search for activity after this date
+        :param updated_before: Search for activity before this date
+        :param override_to_email: Optional override to the destination email
+            address for testing purposes
+        """
+        for digest in self.get_digests(audience, updated_after, updated_before):
+            template_vars = digest.serialize()
+
+            # Check we have something worth sending, and someone to send it to
+            if template_vars["total_annotations"] and (
+                to_email := override_to_email or digest.audience_user.email
+            ):
+                self._mailchimp_service.send_template(
+                    template_name="instructor-email-digest",
+                    sender=self._sender,
+                    recipient=EmailRecipient(
+                        email=to_email, name=digest.audience_user.display_name
+                    ),
+                    template_vars=template_vars,
+                )
+
+    def get_digests(
+        self, audience: List[str], updated_after: datetime, updated_before: datetime
+    ) -> Iterator[Digest]:
+        """
+        Get activity digests for users over a period.
+
+        :param audience: List of H userids to send digests to
+        :param updated_after: Search for activity after this date
+        :param updated_before: Search for activity before this date
+        """
+        annotations = self._h_api.get_annotations(
+            audience, updated_after, updated_before
+        )
```

Comment:
This is kind of annoying, as it bloats our interface a lot. We could accept a list of annotations instead in this and the send email method.

This would mean our tests wouldn't have this sort of thing everywhere:

```
        svc.send_instructor_email_digests(
            audience, sentinel.updated_after, sentinel.updated_before
        )

        h_api.get_annotations.assert_called_once_with(
            audience, sentinel.updated_after, sentinel.updated_before
        )
```

This would complicate the caller a bit, but we only have one caller...